File: mekaneck/virtualbrain/core/types/partition_coord.py
"""
PartitionCoord: Quantum-style categorical coordinates (n, l, m, s).

Implements the partition coordinate system with capacity C(n) = 2n^2.
Maps categorical states to a hierarchical addressing scheme analogous
to atomic quantum numbers but for categorical neural states.
"""

from dataclasses import dataclass
from typing import List, Iterator
import numpy as np


@dataclass(frozen=True)
class PartitionCoord:
    """
    Partition coordinate representing categorical state location.

    Analogous to quantum numbers (n, l, m, s) but for categorical states.
    Provides a complete labeling system for neural partition states.

    Attributes:
        n: Principal quantum number (partition depth), n >= 1
        l: Orbital quantum number (complexity), 0 <= l < n
        m: Magnetic quantum number (orientation), -l <= m <= +l
        s: Spin quantum number (chirality), s in {-0.5, +0.5}

    The capacity at level n is C(n) = 2n^2, giving the sequence:
    C(1)=2, C(2)=8, C(3)=18, C(4)=32, C(5)=50, ...
    """

    n: int
    l: int
    m: int
    s: float

    def __post_init__(self) -> None:
        """Validate quantum number constraints."""
        if self.n < 1:
            raise ValueError(f"n must be >= 1, got {self.n}")
        if not (0 <= self.l < self.n):
            raise ValueError(f"l must be in [0, n), got l={self.l}, n={self.n}")
        if not (-self.l <= self.m <= self.l):
            raise ValueError(f"m must be in [-l, l], got m={self.m}, l={self.l}")
        if self.s not in (-0.5, 0.5):
            raise ValueError(f"s must be +/-0.5, got {self.s}")

    @staticmethod
    def capacity(n: int) -> int:
        """
        Compute capacity at level n.

        C(n) = 2n^2

        This is the total number of distinct categorical states
        at partition depth n.

        Args:
            n: Principal quantum number (partition depth)

        Returns:
            Total number of states at level n
        """
        if n < 1:
            raise ValueError(f"n must be >= 1, got {n}")
        return 2 * n * n

    @staticmethod
    def total_capacity(n_max: int) -> int:
        """
        Compute cumulative capacity up to level n_max.

        Sum_{i=1}^{n_max} C(i) = Sum_{i=1}^{n_max} 2i^2
                               = 2 * n_max(n_max+1)(2n_max+1)/6

        Args:
            n_max: Maximum principal quantum number

        Returns:
            Total number of states up to and including level n_max
        """
        if n_max < 1:
            return 0
        # Using the formula for sum of squares
        return n_max * (n_max + 1) * (2 * n_max + 1) // 3

    @staticmethod
    def density_of_states(n: int) -> int:
        """
        Compute density of states at level n.

        rho(n) = dC/dn = 4n

        Args:
            n: Principal quantum number

        Returns:
            Density of states
        """
        return 4 * n

    def to_linear_index(self) -> int:
        """
        Convert partition coordinate to linear index.

        Enables O(1) array-based lookups by mapping the 4D coordinate
        to a unique integer.

        Returns:
            Linear index for this coordinate
        """
        # States before this n level
        offset = self.total_capacity(self.n - 1)

        # States within this n level, ordered by (l, m, s)
        l_offset = 0
        for l_val in range(self.l):
            l_offset += 2 * (2 * l_val + 1)  # 2*(2l+1) states per l

        m_offset = self.m + self.l  # Map m from [-l, l] to [0, 2l]
        s_offset = 0 if self.s == -0.5 else 1

        return offset + l_offset + 2 * m_offset + s_offset

    @classmethod
    def from_linear_index(cls, index: int) -> "PartitionCoord":
        """
        Reconstruct PartitionCoord from linear index.

        Inverse of to_linear_index().

        Args:
            index: Linear index

        Returns:
            PartitionCoord corresponding to the index
        """
        if index < 0:
            raise ValueError(f"Index must be non-negative, got {index}")

        # Find n level by binary search or iteration
        n = 1
        while cls.total_capacity(n) <= index:
            n += 1

        # Remaining index within n level
        remaining = index - cls.total_capacity(n - 1)

        # Find l by iterating through l values
        l = 0
        while True:
            states_at_l = 2 * (2 * l + 1)
            if remaining < states_at_l:
                break
            remaining -= states_at_l
            l += 1

        # Find m and s from remaining
        m = remaining // 2 - l
        s = -0.5 if remaining % 2 == 0 else 0.5

        return cls(n=n, l=l, m=m, s=s)

    def categorical_distance(self, other: "PartitionCoord") -> float:
        """
        Compute categorical distance D_cat between coordinates.

        D_cat = sqrt((dn)^2 + (dl)^2 + (dm)^2 + (ds)^2)

        This distance is modality-independent - the same distance
        applies regardless of which sensory pathway led to the state.

        Args:
            other: Another partition coordinate

        Returns:
            Categorical distance (dimensionless)
        """
        dn = self.n - other.n
        dl = self.l - other.l
        dm = self.m - other.m
        ds = self.s - other.s

        return float(np.sqrt(dn**2 + dl**2 + dm**2 + ds**2))

    def is_adjacent(self, other: "PartitionCoord") -> bool:
        """
        Check if coordinates are adjacent (D_cat = 1).

        Adjacent states can transition directly without intermediate states.

        Args:
            other: Another partition coordinate

        Returns:
            True if coordinates are adjacent
        """
        return abs(self.categorical_distance(other) - 1.0) < 1e-10

    @classmethod
    def all_coords_at_level(cls, n: int) -> List["PartitionCoord"]:
        """
        Generate all coordinates at partition level n.

        Args:
            n: Principal quantum number

        Returns:
            List of all valid PartitionCoord at this level
        """
        coords = []
        for l in range(n):
            for m in range(-l, l + 1):
                for s in (-0.5, 0.5):
                    coords.append(cls(n=n, l=l, m=m, s=s))
        return coords

    @classmethod
    def iterate_all(cls, n_max: int) -> Iterator["PartitionCoord"]:
        """
        Iterate through all coordinates up to level n_max.

        Args:
            n_max: Maximum principal quantum number

        Yields:
            PartitionCoord in order of linear index
        """
        for n in range(1, n_max + 1):
            for coord in cls.all_coords_at_level(n):
                yield coord

    def energy_level(self, E_max: float = 1.0) -> float:
        """
        Compute energy level for this coordinate.

        E_n = E_max * (n / n_max)^2

        Args:
            E_max: Maximum energy at highest level

        Returns:
            Energy at this partition level
        """
        # Assume n_max = 10 for normalization
        n_max = 10
        return E_max * (self.n / n_max) ** 2

    def __repr__(self) -> str:
        """String representation."""
        s_str = "+1/2" if self.s == 0.5 else "-1/2"
        return f"PartitionCoord(n={self.n}, l={self.l}, m={self.m}, s={s_str})"

File: virtual_brain/crates/vb-core/src/types/partition_coord.rs
//! PartitionCoord: Quantum-style categorical coordinates (n, l, m, s).
//!
//! Implements the partition coordinate system with capacity C(n) = 2n^2.
//! Maps categorical states to a hierarchical addressing scheme analogous
//! to atomic quantum numbers but for categorical neural states.
//!
//! ```text
//! Capacity formula: C(n) = 2n^2
//!   n=1: C=2,  n=2: C=8,  n=3: C=18,  n=4: C=32, ...
//!
//! Constraints:
//!   n >= 1         (principal quantum number)
//!   0 <= l < n     (orbital quantum number)
//!   -l <= m <= +l  (magnetic quantum number)
//!   s in {-0.5, +0.5}  (spin quantum number)
//! ```

use crate::error::PartitionCoordError;
use serde::{Deserialize, Serialize};

/// Spin values representing chirality
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum Spin {
    /// Spin down (-0.5)
    Down,
    /// Spin up (+0.5)
    Up,
}

impl Spin {
    /// Get the numerical value of the spin
    pub fn value(&self) -> f64 {
        match self {
            Spin::Down => -0.5,
            Spin::Up => 0.5,
        }
    }

    /// Create Spin from numerical value
    pub fn from_value(v: f64) -> Result<Self, PartitionCoordError> {
        if (v - (-0.5)).abs() < 1e-10 {
            Ok(Spin::Down)
        } else if (v - 0.5).abs() < 1e-10 {
            Ok(Spin::Up)
        } else {
            Err(PartitionCoordError::InvalidSpin(v))
        }
    }
}

/// Partition coordinate representing categorical state location.
///
/// Analogous to quantum numbers (n, l, m, s) but for categorical states.
/// This type is immutable and hashable, suitable for use as map keys.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct PartitionCoord {
    /// Principal quantum number (partition depth), n >= 1
    n: i32,
    /// Orbital quantum number (complexity), 0 <= l < n
    l: i32,
    /// Magnetic quantum number (orientation), -l <= m <= +l
    m: i32,
    /// Spin quantum number (chirality)
    s: Spin,
}

impl PartitionCoord {
    /// Create a new PartitionCoord with validation.
    pub fn new(n: i32, l: i32, m: i32, s: Spin) -> Result<Self, PartitionCoordError> {
        if n < 1 {
            return Err(PartitionCoordError::InvalidN(n));
        }
        if l < 0 || l >= n {
            return Err(PartitionCoordError::InvalidL(l, n));
        }
        if m < -l || m > l {
            return Err(PartitionCoordError::InvalidM(m, l));
        }
        Ok(Self { n, l, m, s })
    }

    /// Create from f64 spin value (for Python interop).
    pub fn from_components(n: i32, l: i32, m: i32, s_val: f64) -> Result<Self, PartitionCoordError> {
        let s = Spin::from_value(s_val)?;
        Self::new(n, l, m, s)
    }

    // Accessor methods
    pub fn n(&self) -> i32 {
        self.n
    }
    pub fn l(&self) -> i32 {
        self.l
    }
    pub fn m(&self) -> i32 {
        self.m
    }
    pub fn s(&self) -> Spin {
        self.s
    }
    pub fn s_value(&self) -> f64 {
        self.s.value()
    }

    /// Compute capacity at level n: C(n) = 2n^2
    pub fn capacity(n: i32) -> Result<i32, PartitionCoordError> {
        if n < 1 {
            return Err(PartitionCoordError::InvalidN(n));
        }
        Ok(2 * n * n)
    }

    /// Compute cumulative capacity up to level n_max.
    /// Sum_{i=1}^{n_max} 2i^2 = n_max(n_max+1)(2n_max+1)/3
    pub fn total_capacity(n_max: i32) -> i64 {
        if n_max < 1 {
            return 0;
        }
        let n = n_max as i64;
        n * (n + 1) * (2 * n + 1) / 3
    }

    /// Compute density of states: rho(n) = dC/dn = 4n
    pub fn density_of_states(n: i32) -> i32 {
        4 * n
    }

    /// Convert to linear index for O(1) array lookups.
    pub fn to_linear_index(&self) -> i64 {
        let offset = Self::total_capacity(self.n - 1);

        // States within this n level, ordered by (l, m, s)
        let mut l_offset: i64 = 0;
        for l_val in 0..self.l {
            l_offset += 2 * (2 * l_val as i64 + 1); // 2*(2l+1) states per l
        }

        let m_offset = (self.m + self.l) as i64; // Map m from [-l, l] to [0, 2l]
        let s_offset = match self.s {
            Spin::Down => 0,
            Spin::Up => 1,
        };

        offset + l_offset + 2 * m_offset + s_offset
    }

    /// Reconstruct PartitionCoord from linear index.
    pub fn from_linear_index(index: i64) -> Result<Self, PartitionCoordError> {
        if index < 0 {
            return Err(PartitionCoordError::InvalidIndex(index));
        }

        // Find n level
        let mut n = 1;
        while Self::total_capacity(n) <= index {
            n += 1;
        }

        // Remaining index within n level
        let mut remaining = index - Self::total_capacity(n - 1);

        // Find l
        let mut l = 0;
        loop {
            let states_at_l = 2 * (2 * l as i64 + 1);
            if remaining < states_at_l {
                break;
            }
            remaining -= states_at_l;
            l += 1;
        }

        // Find m and s
        let m = (remaining / 2) as i32 - l;
        let s = if remaining % 2 == 0 {
            Spin::Down
        } else {
            Spin::Up
        };

        Self::new(n, l, m, s)
    }

    /// Compute categorical distance D_cat between coordinates.
    /// D_cat = sqrt((dn)^2 + (dl)^2 + (dm)^2 + (ds)^2)
    pub fn categorical_distance(&self, other: &PartitionCoord) -> f64 {
        let dn = (self.n - other.n) as f64;
        let dl = (self.l - other.l) as f64;
        let dm = (self.m - other.m) as f64;
        let ds = self.s.value() - other.s.value();

        (dn * dn + dl * dl + dm * dm + ds * ds).sqrt()
    }

    /// Check if coordinates are adjacent (D_cat = 1).
    pub fn is_adjacent(&self, other: &PartitionCoord) -> bool {
        (self.categorical_distance(other) - 1.0).abs() < 1e-10
    }

    /// Generate all coordinates at partition level n.
    pub fn all_coords_at_level(n: i32) -> Result<Vec<PartitionCoord>, PartitionCoordError> {
        if n < 1 {
            return Err(PartitionCoordError::InvalidN(n));
        }

        let mut coords = Vec::with_capacity(Self::capacity(n)? as usize);
        for l in 0..n {
            for m in -l..=l {
                coords.push(PartitionCoord::new(n, l, m, Spin::Down)?);
                coords.push(PartitionCoord::new(n, l, m, Spin::Up)?);
            }
        }
        Ok(coords)
    }

    /// Iterator over all coordinates up to level n_max.
    pub fn iter_all(n_max: i32) -> impl Iterator<Item = PartitionCoord> {
        (1..=n_max).flat_map(|n| Self::all_coords_at_level(n).unwrap_or_default())
    }

    /// Compute energy level: E_n = E_max * (n / n_max)^2
    pub fn energy_level(&self, e_max: f64, n_max: i32) -> f64 {
        e_max * (self.n as f64 / n_max as f64).powi(2)
    }
}

impl Default for PartitionCoord {
    fn default() -> Self {
        Self::new(1, 0, 0, Spin::Down).unwrap()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_capacity_formula() {
        assert_eq!(PartitionCoord::capacity(1).unwrap(), 2);
        assert_eq!(PartitionCoord::capacity(2).unwrap(), 8);
        assert_eq!(PartitionCoord::capacity(3).unwrap(), 18);
        assert_eq!(PartitionCoord::capacity(4).unwrap(), 32);
        assert_eq!(PartitionCoord::capacity(5).unwrap(), 50);
    }

    #[test]
    fn test_total_capacity() {
        assert_eq!(PartitionCoord::total_capacity(1), 2);
        assert_eq!(PartitionCoord::total_capacity(2), 10); // 2 + 8
        assert_eq!(PartitionCoord::total_capacity(3), 28); // 2 + 8 + 18
    }

    #[test]
    fn test_linear_index_bijection() {
        for idx in 0..100 {
            let coord = PartitionCoord::from_linear_index(idx).unwrap();
            assert_eq!(coord.to_linear_index(), idx);
        }
    }

    #[test]
    fn test_distance_symmetry() {
        let c1 = PartitionCoord::new(2, 1, 0, Spin::Up).unwrap();
        let c2 = PartitionCoord::new(3, 1, 1, Spin::Down).unwrap();
        assert!((c1.categorical_distance(&c2) - c2.categorical_distance(&c1)).abs() < 1e-10);
    }

    #[test]
    fn test_validation() {
        assert!(PartitionCoord::new(0, 0, 0, Spin::Up).is_err());
        assert!(PartitionCoord::new(1, 1, 0, Spin::Up).is_err()); // l >= n
        assert!(PartitionCoord::new(2, 1, 2, Spin::Up).is_err()); // m > l
    }
}
